pkg/admission/pod_eviction.go

```diff
@@ -0,0 +1,464 @@
+	}
+
+	// evicting a terminal pod should result in direct deletion of pod as it already caused disruption by the time we are evicting.
+	if canIgnorePDB(pod) {
+		level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logAllow, errPodNotReady))
+		return request.allowWithWarning(errPodNotReady)
+	}
+
+	pdbConfig, err := pdbCache.Find(pod)
+	if err != nil {
+		level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - %v", logDeny, err))
+		return request.denyWithReason(err.Error(), httpStatusMisconfiguration)
+	} else if pdbConfig == nil {
+		level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logAllow, errNoPdbForPod))
+		return request.allowWithWarning(errNoPdbForPod)
+	}
+
+	// get the StatefulSet who manages this pod
+	if sts, err = request.client.owner(pod); err != nil {
+		level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - unable to find pod owner", logAllow), "err", err)
+		return request.allowWithWarning(err.Error())
+	}
+
+	// ie ingester-zone-a
+	request.log.SetSpanAndLogTag("owner", sts.Name)
+
+	lock := findLock(pdbConfig.Name())
+	lock.Lock()
+	defer lock.Unlock()
+
+	// the number of allowed unavailable pods in other zones.
+	maxUnavailable := pdbConfig.MaxUnavailablePods(sts)
+	if maxUnavailable == 0 {
+		level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logDeny, errMaxUnavailableIs0))
+		return request.denyWithReason(errMaxUnavailableIs0, httpStatusPdbDisabled)
+	}
+
+	// Find the other StatefulSets which span all zones.
+	// Assumption - each StatefulSet manages all the pods for a single zone. This list of StatefulSets covers all zones.
+	// During a migration it may be possible to break this assumption, but during a migration the maxUnavailable will be set to 0 and the eviction request will be denied.
+	otherStatefulSets, err := request.client.findRelatedStatefulSets(sts, pdbConfig.StsSelector())
+	if err != nil || otherStatefulSets == nil || len(otherStatefulSets.Items) <= 1 {
+		level.Error(request.log).Log(logMsg, "unable to find related stateful sets - continuing with single zone")
+		otherStatefulSets = &appsv1.StatefulSetList{Items: []appsv1.StatefulSet{*sts}}
+	}
+
+	// this is the partition of the pod being evicted - for classic zones the partition will be an empty string
+	partition, err := pdbConfig.PodPartition(pod)
+	if err != nil {
+		return request.denyWithReason(err.Error(), httpStatusMisconfiguration)
+	}
+
+	// include the partition we are considering
+	if len(partition) > 0 {
+		request.log.SetSpanAndLogTag("partition", partition)
+	}
+
+	// include the number of zones we will be considering
+	request.log.SetSpanAndLogTag("zones", len(otherStatefulSets.Items))
+
+	var pdb *pdbValidator
+
+	if len(otherStatefulSets.Items) == 1 {
+		// single zone mode - we include the pod being evicted into the calculation and the maxAvailable is applied to the single zone / StatefulSet
+		pdb = makePdbValidatorSingleZone(&otherStatefulSets.Items[0])
```

Comment:
A single zone seems like a misconfiguration - in the interests of simplifying things, how about we just reject the eviction request if there is a ZPDB but only one matching StatefulSet?